rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,533 @@
+  log.Error("Corruption Detected: %s", d.String())
+
+  // The call back could be used to page or signal the operator in some way.
+
+  if d.Kind == corruption.KindWAL {
+    // Signal partition shutdown, we lost transactions
+    return errors.New("Detected WAL corruption, database shutdown")
+  }
+
+  if d.Kind == corruption.KindSST {
+    // Signal bit rot, rebuild the partition from replicas, but allow reads to continue
+    if err := signalCorruption(d); err != nil {
+      log.Error("while signaling: %s", err)
+    }
+  }
+  // Else we return no error, and the database remains available.
+  return nil
+}
+```
+##### Rust Example Usage
+```rust
+let opts = DBOptions {
+  on_corruption: Some(Box::new(|details: CorruptionDetails| -> Result<(), CorruptionError> {
+    // Log corruption details
+    log::error!("Corruption Detected: {}", details);
+
+    // Handle different kinds of corruption
+    match details.kind {
+      CorruptionKind::WAL => {
+        // Signal partition shutdown due to WAL corruption
+        return Err(CorruptionError { details: details.clone() });
+      }
+      CorruptionKind::SST => {
+        // Signal bit rot, rebuild from replicas, allow reads to continue
+        if let Err(err) = signal_corruption(&details) {
+          log::error!("Error while signaling corruption: {}", err);
+        }
+        Ok(())
+      }
+      _ => {
+        // For other kinds of corruption, return no error and keep the database available
+        Ok(())
+      }
+    }
+  })),
+  ..Default::default()
+};
+
+let db = Db::open_with_opts("path/to/db", opts, object_store).await;
+```
+
+If `DBOptions.on_corruption` is not defined, `DBOptions.on_corruption` will default to 
+a function which both logs the error and always returns an error, thus making the 
+database unavailable by default.
+
+#### Handling WAL Corruption during Recovery
+It is reasonable for a developer to assume that once data is confirmed to be
+written to the Write-Ahead Log (WAL), the data or transaction is considered
+durably written to storage. However, as discussed in the "Corruption Review"
+section of this document, this assumption is only theoretical. In practice,
+corruption may only be discovered when recovering from a system failure or
+crash, which is when WAL entries are most critical.
+
+It is important to note that during normal LSM operation, the movement of
+entries from the WAL to the SSTable occurs in memory. The WAL is only used as a
+recovery mechanism in case of a catastrophic failure of the LSM. Therefore, the
+loss of committed WAL data can only occur when both of the following conditions are
+true.
+
+1. The LSM application exits before compaction can complete.
+2. Corruption of the WAL occurs.
+
+Although the probability of both conditions being true is very low, especially
+in the context of durable object storage, it is not impossible.
+
+To address this, I propose that the `DB::open()` function perform WAL
```

Comment:
I believe `replay_wal` in SlateDB (rust) calls `replay_wal` and is already a synchronous operation. So, we should be good here.